### PR TITLE
feat(mito): Add WriteCache struct and write SSTs to write cache

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements a skeleton for the `WriteCache` that serves as a write-through cache for object stores.

### WriteCache
The `WriteCache` has a local store to cache files.
```rust
pub struct WriteCache {
    /// Local file cache.
    file_cache: FileCacheRef,
    /// Object store manager.
    object_store_manager: ObjectStoreManagerRef,
}
```

The `WriteCache` has a `write_and_upload_sst()` method to write files to the cache (a local store) and upload them to object stores.
```rust
pub(crate) async fn write_and_upload_sst(
    &self,
    request: SstUploadRequest,
    write_opts: &WriteOptions,
) -> Result<Option<SstInfo>>
```

Upload can be sync or async. We will implement the sync version first.

The `recover()` method will recover the underlying file cache.
```rust
pub async fn recover(&self) -> Result<()>;
```

The `AccessLayer` hides the details of writing SSTs to the write cache.
```rust
impl AccessLayer {
    pub(crate) async fn write_sst(
        &self,
        request: SstWriteRequest,
        write_opts: &WriteOptions,
    ) -> Result<Option<SstInfo>> {
        // ...
    }
}
```

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/2965

### CODE REVIEW

**Review by WenyXu — APPROVED** [review]

File: src/mito2/src/test_util/scheduler_util.rs [hunk 1]
```diff
@@ -65,7 +66,11 @@ impl SchedulerEnv {
     ) -> CompactionScheduler {
         let scheduler = self.get_scheduler();
 
-        CompactionScheduler::new(scheduler, request_sender)
+        CompactionScheduler::new(
+            scheduler,
+            request_sender,
+            Arc::new(CacheManager::new(0, 0, 0)),
```
Comment (killme2008) [inline src/mito2/src/test_util/scheduler_util.rs:72]: We'd better test the mito engine with the valid cache manager when it's ready.

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/mito2/src/cache/write_cache.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+use crate::request::WorkerRequest;
+use crate::sst::file::{FileId, FileMeta, Level};
+use crate::sst::parquet::writer::ParquetWriter;
+use crate::sst::parquet::WriteOptions;
+use crate::wal::EntryId;
+
+/// A cache for uploading files to remote object stores.
+///
+/// It keeps files in local disk and then sends files to object stores.
+pub(crate) struct WriteCache {
```
Comment (zhongzc) [inline src/mito2/src/cache/write_cache.rs:37]: I feel curious about how `WriteCache` and `UploadPartWriter` will interact in the future

File: src/mito2/src/cache/write_cache.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+/// A remote write request to upload files.
+pub(crate) struct Upload {
+    /// Parts to upload.
+    pub(crate) parts: Vec<UploadPart>,
+}
+
+/// Metadata of SSTs to upload together.
+pub(crate) struct UploadPart {
```
Comment (zhongzc) [inline src/mito2/src/cache/write_cache.rs:75]: If the `UploadPart` only serves for SSTs, can I still have the chance to write files of other types in the parquet writer?

File: src/mito2/src/cache/write_cache.rs (hunk 2 above)
Comment (evenyag) [reply src/mito2/src/cache/write_cache.rs:37]: You can refer to the POC branch.

https://github.com/evenyag/greptimedb/blob/14a325f2275c792d0f495b21b1ca1e1ef37f844d/src/mito2/src/flush.rs#L359-L366

File: src/mito2/src/cache/write_cache.rs (hunk 3 above)
Comment (evenyag) [reply src/mito2/src/cache/write_cache.rs:75]: Good question. Now it only serves for SSTs, we might store the file type in the `FileMeta` to support other files.

**Review by zhongzc — APPROVED** [review]

**Comment by evenyag** [comment]
After some experiments, I plan to simplify the implementation of the write-through cache and upload files to object stores synchronously. Then we can remove `UploadPart` related components. I'll get back soon.

File: src/mito2/src/test_util/scheduler_util.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/test_util/scheduler_util.rs:72]: This util only mocks a scheduler for scheduler tests. The scheduler itself doesn't care about the cache manager so it's fine to provide an empty cache here.

**Comment by evenyag** [comment]
I refactored the `AccessLayer` and `WriteCache`:
- https://github.com/GreptimeTeam/greptimedb/pull/2999/commits/02785134ca579e005334f0539b1bbd410db2d765 refactors the `AccessLayer` to hide the details of writing SST files.
- https://github.com/GreptimeTeam/greptimedb/pull/2999/commits/9f03e6746f681c960e6bb5b2988b124bd2d4c1a3 writes the SST to the write cache, though it doesn't go through the local file cache.

**Review by zhongzc — APPROVED** [review]

File: src/mito2/src/cache/write_cache.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        // TODO(yingwen): Write to the local store and then upload.
+        // Now we write to the remote and ignore local cache.
+        let mut writer =
+            ParquetWriter::new(request.upload_path, request.metadata, request.remote_store);
+        writer.write_all(request.source, write_opts).await
+    }
+}
+
+/// Request to write and upload a SST.
+pub(crate) struct SstUploadRequest {
```
Comment (killme2008) [inline src/mito2/src/cache/write_cache.rs:77]: If we already set the struct is `pub(crate)`, looks like we don't need to set every field to be `pub(crate)`.

Comment (evenyag) [reply src/mito2/src/cache/write_cache.rs:77]: No, fields are still private without `pub(crate)`.

Comment (killme2008) [reply src/mito2/src/cache/write_cache.rs:77]: I meant that they can be just `pub` instead of `pub(crate)`.

Comment (evenyag) [reply src/mito2/src/cache/write_cache.rs]: I pub `write_and_upload_sst()` and this struct.

**Pull request merged**